FEATURE: Ignoring users from the preferences page in a cleaner way

khalilovcmded opened the pull request:
## UI

![Revised UI](https://user-images.githubusercontent.com/45508821/56358735-72005c80-61d7-11e9-95d2-a9892998dacc.gif)

Comparing it to the PM participants:

![Naka-Karuizawa_trip__-_Discourse](https://user-images.githubusercontent.com/45508821/56364951-b4319a00-61e7-11e9-9423-3ae522381ea9.png)

tgxworld commented:
I'm wondering if a list will be better here for the ignored users. Instead of trying to fit everything in an input field, we can list the users alongside the duration that the user will be ignored for.

coding-horror commented:
We don’t need to list duration though, so I think a simpler approach is
better.
On Sun, Mar 31, 2019 at 8:57 PM Guo Xiang Tan <notifications@github.com>
wrote:

> I'm wondering if a list will be better here for the ignored users. Instead
> of trying to fit everything in an input field, we can list the users
> alongside the duration that the user will be ignored for.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/discourse/discourse/pull/7289#issuecomment-478410173>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABc7VdPQM4RnDJWolDwi5P8h4oIgHYnIks5vcWeZgaJpZM4cSyyI>
> .
>

tgxworld commented:
Ah ok :) I thought the duration would be helpful.

ZogStriP commented on app/assets/javascripts/discourse/components/ignored-user-list-item.js.es6:
```diff
@@ -0,0 +1,14 @@
+export default Ember.Component.extend({
+  classNames: ['item'],
+  tagName: "div",
+  router: Ember.inject.service(),
```
Is that needed?

ZogStriP commented on app/assets/javascripts/discourse/components/ignored-user-list-item.js.es6:
```diff
@@ -0,0 +1,14 @@
+/* You might be looking for navigation-item. */
```
What is that comment for?

ZogStriP commented on app/assets/javascripts/discourse/components/ignored-user-list.js.es6:
```diff
@@ -0,0 +1,33 @@
+      });
+    },
+    newItem() {
+      const controller = showModal("ignore-duration-with-username", {
```
`controller` is too generic and should be rarely used.

```suggestion
      const modal = showModal("ignore-duration-with-username", {
```

ZogStriP commented on app/assets/javascripts/discourse/components/ignored-user-list.js.es6:
```diff
@@ -0,0 +1,33 @@
+        onSuccess: (username) => {
+          this.get("items").addObject(username);
+        },
+        onClose: () => {
```
Is that needed?

ZogStriP commented on app/assets/javascripts/discourse/templates/components/ignored-user-list.hbs:
```diff
@@ -0,0 +1,7 @@
+<div class="ac-wrap clearfix" style="width: 480px;">
```
No inline styling ;)

ZogStriP commented on app/assets/javascripts/discourse/templates/modal/ignore-duration-with-username.hbs:
```diff
@@ -0,0 +1,26 @@
+{{#d-modal-body title="user.user_notifications.ignore_duration_title" autoFocus="false"}}
+  <div class="controls tracking-controls">
+    <label>{{d-icon "eye-slash" class="icon"}} Username</label>
```
`Username` should be localized

khalilovcmded commented:
@awesomerobot Can you please take a look at this proposed pattern and let me know what you think?

My concern here is that I am introducing a new UX pattern that might not be very consistent or transferable to other parts of Discourse. I wonder if there are any tweaks we could do to improve this. Also, an acceptable tweak is to leave the current version in production as it is 😉 

@jjaffeux made a good point about having a quick chat with you.

coding-horror commented:
if we can re-use existing code (e.g. the PM participants) I am for it, otherwise simple is better.. no need to gold plate this feature, we just need it to exist, not be perfect

khalilovcmded commented:
@awesomerobot can you please take a look at [this](https://github.com/discourse/discourse/pull/7289#issue-265787516)? 

It is not much code and maps to the behaviour we have in the PM participants.

khalilovcmded commented:
@awesomerobot that's a good point Kris 👍 I have changed it:

![Profile_-_tarek_-_Discourse](https://user-images.githubusercontent.com/45508821/56371372-e990b480-61f4-11e9-8bcb-05c3ce1a5821.png)

khalilovcmded commented:
I have just added one more detail, when there are no ignored users, it looks strange as there is a void between the `ignored` title and the space below it.

It used to look like this:

![Profile_-_tarek_-_Discourse](https://user-images.githubusercontent.com/45508821/56416703-96b50c80-6289-11e9-87c0-429b668ec584.png)

I changed it to look like this:

![Profile_-_tarek_-_Discourse](https://user-images.githubusercontent.com/45508821/56416682-8c930e00-6289-11e9-8aac-a2f4f8c250ba.png)

ZogStriP commented:
I wonder if we should instead move the gray description next to the title

![image](https://user-images.githubusercontent.com/362783/56417100-24e1c080-6293-11e9-9847-0142141e8bdb.png)

khalilovcmded commented:
That looks or will look very strange, I don't think it is good.

It also breaks the pattern that we established: `Title -> input -> instruction` like:
![Profile_-_tarek_-_Discourse](https://user-images.githubusercontent.com/45508821/56417386-a7668200-628b-11e9-9dc8-ead1b8b09ace.png)

khalilovcmded commented:
@ZogStriP @awesomerobot shall we try and merge this?

ZogStriP commented on app/assets/javascripts/discourse/components/ignored-user-list-item.js.es6:
```diff
@@ -0,0 +1,9 @@
+  items: null,
+  actions: {
+    removeItem(item) {
+      this.get("onRemoveItem")(item);
```
Why isn't this directly sending the proper action?

```suggestion
      this.send("removeItem", item);
```

ZogStriP commented on app/assets/javascripts/discourse/components/ignored-user-list.js.es6:
```diff
@@ -0,0 +1,29 @@
+        model: this.get("model")
+      });
+      modal.setProperties({
+        onSuccess: username => {
```
Why isn't this an action defined in the `ignore-duration-with-username` controller?

khalilovcmded replied:
The `ignore-duration-with-username` controller / modal doesn't have access to `items`. The idea here is that the controller doesn't need to be aware of `items` since its responsibility is to `take care of the ignore user action` and on its success `we` (the `ignored-user-list`) are responsible of what to do next.

ZogStriP commented on app/assets/javascripts/discourse/controllers/ignore-duration-with-username.js.es6:
```diff
@@ -0,0 +1,31 @@
+        user
+          .updateNotificationLevel("ignore", this.get("ignoredUntil"))
+          .then(() => {
+            this.get("onSuccess")(this.get("ignoredUsername"));
```
Why isn't this sending an action instead?

khalilovcmded replied:
Communication from a controller to its route can be done using `send`, but here we are communicating from a controller to its calling component.

khalilovcmded replied:
We would / could have used `sendAction` but that's deprecated.

khalilovcmded commented on app/assets/javascripts/discourse/components/ignored-user-list-item.js.es6:
```diff
@@ -0,0 +1,9 @@
+  items: null,
+  actions: {
+    removeItem(item) {
+      this.send("removeItem", item);
```
Just realised, we can't actually do that, we are communicating back to the parent component. We should keep it as it originally was.

jjaffeux commented on app/assets/javascripts/discourse/templates/components/ignored-user-list-item.hbs:
```diff
@@ -0,0 +1,4 @@
+<div class="ignored-user-list-item">
+  <span class="ignored-user-name">{{item}}</span>
+  <a class="remove-ignored-user no-text btn-icon btn" {{action "removeItem" item}}>{{d-icon "times"}}</a>
```
```suggestion
  {{d-button class="remove-ignored-user no-text btn-icon" action=(action "removeItem" item) icon="times"}}
```

jjaffeux commented on app/assets/javascripts/discourse/templates/components/ignored-user-list.hbs:
```diff
@@ -0,0 +1,11 @@
+  {{/if}}
+</div>
+<div class="instructions">{{i18n 'user.ignored_users_instructions'}}</div>
+<div>{{#d-button action=(action "newItem") class="btn-default" icon="plus"}}Add ...{{/d-button}}</div>
```
Should be localised

jjaffeux commented on app/assets/stylesheets/common/components/ignored-user-list.scss:
```diff
@@ -0,0 +1,37 @@
+  padding: 0;
+  display: flex;
+  align-items: center;
+  height: 26px;
```
can we avoid this kind of arbitrary height ?

jjaffeux commented on app/assets/javascripts/discourse/components/ignored-user-list.js.es6:
```diff
@@ -0,0 +1,29 @@
+export default Ember.Component.extend({
+  item: null,
+  actions: {
+    removeItem(item) {
```
removeItem and newItem seem too generic to me

jjaffeux commented on app/assets/javascripts/discourse/components/ignored-user-list.js.es6:
```diff
@@ -0,0 +1,29 @@
+        model: this.get("model")
+      });
+      modal.setProperties({
+        onSuccess: username => {
```
onSuccess seems to generic to me, what about something like: `onUserIgnored` ?

khalilovcmded replied:
Much better naming!

jjaffeux commented on app/assets/javascripts/discourse/components/ignored-user-list.js.es6:
```diff
@@ -0,0 +1,29 @@
+      });
+    },
+    newItem() {
+      const modal = showModal("ignore-duration-with-username", {
```
depending on your needs you could rewrite it as:

```
const modal = showModal("ignore-duration-with-username", {
  model: {
    betterModelName: this.get("model"),
    onUserIgnored: username => {
      this.get("items").addObject(username);
    }
  }
});
```

khalilovcmded replied:
I am not sure you can do that with the `showModal`

jjaffeux commented on app/assets/javascripts/discourse/controllers/ignore-duration-with-username.js.es6:
```diff
@@ -0,0 +1,31 @@
+        user
+          .updateNotificationLevel("ignore", this.get("ignoredUntil"))
+          .then(() => {
+            this.get("onSuccess")(this.get("ignoredUsername"));
```
if you are sure onSuccess is always defined here, you can do: `this.onSucess(this.get("ignoredUsername")`

The get here won’t protect you from undefined object as you call a function on it:

```
this.get("unknown")()
// Uncaught TypeError: this.get(...) is not a function
```

khalilovcmded replied:
I think it is less to do with `protection` from calling an undefined and more to `standari`ing` the way I access the component's properties, though I think it is okay to change it here 👍

jjaffeux commented on app/assets/javascripts/discourse/templates/preferences/users.hbs:
```diff
@@ -1,20 +1,16 @@
 <div class="control-group">
   <div class="controls tracking-controls">
-    <label>{{d-icon "d-muted" class="``icon"}} {{i18n 'user.muted_users'}}</label>
+    <label>{{d-icon "d-muted" class="icon"}} {{i18n 'user.muted_users'}}</label>
```
```suggestion
    <label>
      {{d-icon "d-muted" class="icon"}}
      <span>{{i18n 'user.muted_users'}}</span>
    </label>
```

jjaffeux commented on app/assets/javascripts/discourse/templates/components/ignored-user-list.hbs:
```diff
@@ -0,0 +1,11 @@
+<div class="ignored-list">
+  {{#if items}}
+    {{#each items as |item|}}
+      {{ignored-user-list-item item=item onRemoveItem=(action "removeItem")}}
```
onRemoveItem feels too generic

jjaffeux commented on app/assets/javascripts/discourse/templates/components/ignored-user-list.hbs:
```diff
@@ -0,0 +1,11 @@
+<div class="ignored-list">
+  {{#if items}}
```
each already has this behavior:

```
{{#each unknownItems as |item|}}
{{else}}
    will print this message
{{/each}}
```

jjaffeux commented:
I didn't put it each time, but generally speaking for a feature with a clear usage and wording, if we can avoid "item" I think it's better.

khalilovcmded commented:
> I didn't put it each time, but generally speaking for a feature with a clear usage and wording, if we can avoid "item" I think it's better.

I find it hard to agree. I personally adopted this convention through looking & working on other Ember projects that seem to agree on this pattern. For example, **Ghost's** Ember project, you will see that mostly children components to some an array models are usually postfixed with `item`:

![File_Finder](https://user-images.githubusercontent.com/45508821/56662111-2815eb80-669b-11e9-9f28-3b85313ff508.png)

khalilovcmded commented:
One further iteration, I feel this is better compartmentalisation, the `save changes` button is actually only used for muted users now:

![Profile_-_tarek_-_Discourse](https://user-images.githubusercontent.com/45508821/56663251-8f349f80-669d-11e9-81c9-c34ee327c8f5.png)

jjaffeux commented:
To be clear I was not talking of files naming which I agree is fine. It's more in the actions and variables naming.

jjaffeux commented:
@venarius both tests failures seems legit and related to the commit

jjaffeux commented:
@ZogStriP looks good to me 👍 

👏 @khalilovcmded
